lib/activerecord-import/import.rb

```diff
@@ -24,48 +24,40 @@ def initialize(name, index)
   end
 
   class Validator
-    def initialize(validators, options = {})
-      @validators = validators
-      @options    = options
+    def initialize(options = {})
+      @options = options
     end
 
     def valid_model?(model)
       validation_context = @options[:validate_with_context]
       validation_context ||= (model.new_record? ? :create : :update)
 
-      current_context = model.send(:validation_context)
-      model.send(:validation_context=, validation_context)
-      model.errors.clear
+      validate_callbacks_orig = model._validate_callbacks
 
-      validation_proc = lambda do
-        @validators.each do |v|
-          if validation_context == v.options.fetch(:on, validation_context)
-            v.validate(model) if validate?(v, model)
-          end
-        end
+      validate_callbacks = model._validate_callbacks.dup
+      validate_callbacks.each do |callback|
+        validate_callbacks.delete(callback) if callback.raw_filter.is_a? ActiveRecord::Validations::UniquenessValidator
       end
 
-      if model.respond_to?(:run_validation_callbacks)
-        model.send(:_run_validation_callbacks, &validation_proc)
-      else
-        model.send(:run_callbacks, :validation, &validation_proc)
+      begin
+        set_validate_callbacks(model, validate_callbacks)
+        model.valid?(validation_context)
+      ensure
+        set_validate_callbacks(model, validate_callbacks_orig)
       end
-
-      model.send(:validation_context=, current_context)
-      model.errors.empty?
     end
 
-    def validate?(validator, model)
-      evaluate = lambda do |condition|
-        case condition
-        when String then model.instance_eval(condition)
-        when Symbol then model.send(condition)
-        when Proc then model.instance_eval(&condition)
-        end
-      end
+    private
 
-      Array(validator.options[:if]).map(&evaluate).compact.all? &&
-        !Array(validator.options[:unless]).map(&evaluate).compact.any?
+    def set_validate_callbacks(model, validate_callbacks)
+      if model.respond_to?(:_validate_callbacks=) # ActiveRecord < 4.1
+        model._validate_callbacks = validate_callbacks
+        model.class.__reset_runner(:validate)
+      elsif model.class.respond_to?(:_validate_callbacks=) # ActiveRecord < 5.1
+        model.class._validate_callbacks = validate_callbacks
```

Comment:
Yeah, you're right :frowning_face:

Let me rework this so it doesn't change anything at the class level. (Note that modifying the model *instance* isn't thread-safe either, but ActiveRecord is already thread-unsafe in this regard, so it's probably fine to assume that the same instance won't be simultaneously accessed from multiple threads.)